File: Algorithms/Python/countingsort.py
# Python program for counting sort, contributed by Gokul S
#Time Complexity: O(n+k) where n is the number of elements in input array and k is the range of input. 
#Auxiliary Space Complexity: O(n+k)
  
def countSort(arr): 
  
    # output arr
    output = [0 for i in range(len(arr))] 
  
    # to store count of each char
    count = [0 for i in range(256)] 
  
    # store resulting ans
    ans = ["" for _ in arr] 
  
    # Storing count of each char
    for i in arr: 
        count[ord(i)] += 1
  
    # Change count[i] so that count[i] now contains actual position of this character in output arr
    for i in range(256): 
        count[i] += count[i-1] 
 
    for i in range(len(arr)): 
        output[count[ord(arr[i])]-1] = arr[i] 
        count[ord(arr[i])] -= 1
  
    # Copy the output arr so that arr now contains sorted characters 
    for i in range(len(arr)): 
        ans[i] = output[i] 
    return ans  
  
# Driver
arr = "welcometohacktoberfest"
ans = countSort(arr) 
print("Sorted character array is % s" %("".join(ans)))

File: Algorithms/Java/countingSort.java
// Java implementation of Counting Sort 
class CountingSort { 
	public static void sort(char arr[]) 
	{ 
		int n = arr.length; 
		char output[] = new char[n]; 
		int count[] = new int[256]; 
		for (int i = 0; i < 256; ++i) 
			count[i] = 0; 

		for (int i = 0; i < n; ++i) 
			++count[arr[i]]; 

		for (int i = 1; i <= 255; ++i) 
			count[i] += count[i - 1]; 

		for (int i = n - 1; i >= 0; i--) { 
			output[count[arr[i]] - 1] = arr[i]; 
			--count[arr[i]]; 
		} 
		for (int i = 0; i < n; ++i) 
			arr[i] = output[i]; 
	} 
	public static void main(String args[]) 
	{ 
		char arr[] = {'h','a','c','k','t','o','b','e','r','f','e','s','t'}; 
		System.out.println("character array before sorting: "); 
		for (int i = 0; i < arr.length; ++i) 
			System.out.print(arr[i]);
                System.out.println();
                sort(arr); 
		System.out.println("Sorted character array is "); 
		for (int i = 0; i < arr.length; ++i) 
			System.out.print(arr[i]);
                System.out.println(); 
	} 
} 

